### ISSUE TITLE
Commbench takes over 30 mins to run on Golub (with both ucx, verbs) 

### ISSUE THREAD

**Issue by nitbhat** [UIUC-PPL/charm#3257]
There is some weird performance issue on Golub which causes commbench to take over 30-35 mins to run. This is delaying ucx and verbs autobuild runs. (both smp and nonsmp). 

The same test takes about 5s or so to run on Golub and Frontera.

**Comment by rbuch** [comment]
> There is some weird performance issue on Golub which causes commbench to take over 30-35 mins to run. This is delaying ucx and verbs autobuild runs. (both smp and nonsmp).
> 
> The same test takes about 5s or so to run on Golub and Frontera.

I assume you mean it takes about 5s or so to run on some other machine and Frontera, not Golub and Frontera, right?

**Comment by nitbhat** [comment]
> > There is some weird performance issue on Golub which causes commbench to take over 30-35 mins to run. This is delaying ucx and verbs autobuild runs. (both smp and nonsmp).
> > The same test takes about 5s or so to run on Golub and Frontera.
> 
> I assume you mean it takes about 5s or so to run on some other machine and Frontera, not Golub and Frontera, right?

Sorry, that was typo. I meant to write that it takes 5s on Cori and Frontera. (and not Golub, where it takes about 30 mins).

**Comment by nitbhat** [comment]
As suggested by @matthiasdiener, I dug into the autobuild logs and found that the commbench's run on Golub on Sept 16, 2020 took about 6s (http://charm.cs.illinois.edu/autobuild/old.2020_09_16__01_04/ucx-linux-x86_64.txt). 

Since then, I see that for a long period the tests haven't successfully run (and when they did they took about 30 odd mins). 

Currently, we're seeing some build issues on Golub because of the unavailability of verbs/ucx headers. I've reached out to the sysadmins to see what's causing them. Once they're fixed, I can check if there's a particular charm commit causing the slowdown. (or if it's being caused by some environment change on Golub that happened around that time frame)

**Comment by matthiasdiener** [comment]
This is a test that took ~27 mins: http://charm.cs.illinois.edu/autobuild/old.2021_01_01__01_15/ucx-linux-x86_64.txt

A diff of the output for the commbench looks like this:

```patch
--- fast
+++ slow
@@ -1,82 +1,82 @@
-make[3]: Entering directory `/scratch/users/nbhat4/autobuild/ucx/charm/ucx-linux-x86_64/benchmarks/converse/commbench'
-../../../bin/testrun  ./pgm +p4  
+make[2]: Entering directory `/scratch/users/nbhat4/autobuild/ucx/charm/ucx-linux-x86_64/benchmarks/converse/commbench'
+../../../bin/testrun  ./commbench +p4  
 
-Running on 4 processors:  ./pgm 
-./charmrun: line 152: mvapich2-start-mpd: command not found
+Running on 4 processors:  ./commbench 
+charmrun>  /usr/bin/setarch x86_64 -R  mpirun -np 4  ./commbench 
 Charm++> Running in non-SMP mode: 4 processes (PEs)
-Converse/Charm++ Commit ID: dc40786
+Converse/Charm++ Commit ID: e32c995
 Charm++ built with internal error checking enabled.
 Do not use for performance benchmarking (build without --enable-error-checking to do so).
 Isomalloc> Synchronized global address space.
-proc module initproc module init
 proc module init
-
 proc module init
-Charm++> Running on 2 hosts (1 sockets x 3 cores x 1 PUs = 3-way SMP)
-Charm++> cpu topology info is gathered in 0.002 seconds.
+proc module init
+proc module init
+Charm++> Running on 1 hosts
+Charm++> cpu topology info is gathered in 0.070 seconds.
 [memoryAccess] initiated
 Single core malloc/free/traversal performance numbers
 Performance number of clubbed malloc-traversal-free
 Size   Iterations  Malloc  Write   Read    Free
-1  1000    5.398100E-05    9.620000E-07    1.762000E-06    2.217200E-05
-2  1000    5.375800E-05    3.946000E-06    6.273000E-06    2.221300E-05
-4  1000    6.323600E-05    8.076000E-06    1.484400E-05    2.240200E-05
-8  1000    8.064700E-05    1.672200E-05    3.151400E-05    2.248500E-05
-16 1000    6.512200E-05    3.902400E-05    6.832800E-05    2.950100E-05
-32 1000    7.417500E-05    8.703500E-05    1.419250E-04    2.993800E-05
-64 1000    1.637140E-04    1.983520E-04    2.789400E-04    3.038600E-05
-128    1000    2.924730E-04    3.510310E-04    4.337710E-04    2.441000E-05
-256    1000    4.696970E-04    6.574960E-04    8.656170E-04    2.502700E-05
-512    1000    9.009340E-04    1.263822E-03    1.786493E-03    2.843900E-05
-1024   1000    9.725630E-04    2.955682E-03    3.461533E-03    4.898500E-05
-2048   1000    1.031255E-03    5.178679E-03    6.934119E-03    8.959620E-04
-4096   1000    1.583965E-03    1.103826E-02    1.401504E-02    1.919953E-03
-8192   1000    1.798763E-03    2.237240E-02    2.792006E-02    3.575979E-03
-16384  1000    2.327002E-03    4.511277E-02    5.573226E-02    6.740397E-03
-Time taken by random malloc-traversal-free benchmark with following commutative distribution: malloc 1, write 5, read 9, free 10, max malloc size 16384, length of table 1000, number of ops 1000 is 3.950772E-02.
+1  1000    3.564900E-05    8.820000E-07    1.351000E-06    1.691200E-05
+2  1000    3.675200E-05    2.973000E-06    4.921000E-06    1.683900E-05
+4  1000    4.134400E-05    5.983000E-06    1.136500E-05    1.699900E-05
+8  1000    4.266600E-05    1.281800E-05    2.403800E-05    1.700100E-05
+16 1000    4.009500E-05    2.919800E-05    4.942500E-05    2.215300E-05
+32 1000    6.329500E-05    6.556200E-05    1.136850E-04    2.289800E-05
+64 1000    1.112160E-04    1.505350E-04    2.069040E-04    2.313400E-05
+128    1000    2.537400E-04    2.951290E-04    4.115240E-04    2.359400E-05
+256    1000    4.565750E-04    5.883600E-04    8.102170E-04    2.372500E-05
+512    1000    8.745550E-04    1.183066E-03    1.614334E-03    2.616700E-05
+1024   1000    9.054940E-04    2.374671E-03    2.612279E-02    2.941400E-05
+2048   1000    9.552450E-04    4.753395E-03    2.743384E-02    1.696007E-02
+4096   1000    1.389822E-03    4.564581E-02    3.393383E-02    2.278250E-02
+8192   1000    1.680703E-03    8.729673E-02    1.072845E-01    1.041146E-02
+16384  1000    2.282818E-03    1.488630E-01    2.222197E-01    3.000417E-02
+Time taken by random malloc-traversal-free benchmark with following commutative distribution: malloc 1, write 5, read 9, free 10, max malloc size 16384, length of table 1000, number of ops 1000 is 1.518634E-01.
 Multi core malloc/free/traversal performance numbers
 Performance number of clubbed malloc-traversal-free
 Size   Iterations  Malloc  Write   Read    Free
-1  1000    2.985900E-05    8.450000E-07    1.448000E-06    1.782100E-05
-2  1000    3.134500E-05    3.217000E-06    5.110000E-06    1.801200E-05
-4  1000    3.634000E-05    6.405000E-06    1.207600E-05    1.808100E-05
-8  1000    4.104900E-05    1.420100E-05    2.652000E-05    1.878900E-05
-16 1000    5.223900E-05    3.280800E-05    6.168500E-05    2.471200E-05
-32 1000    4.892100E-05    7.022400E-05    1.144720E-04    2.423000E-05
-64 1000    1.029160E-04    1.599810E-04    2.300070E-04    2.452800E-05
-128    1000    1.844810E-04    3.131350E-04    4.377780E-04    2.517100E-05
-256    1000    4.684060E-04    6.271160E-04    8.916780E-04    2.569500E-05
-512    1000    8.853040E-04    1.257153E-03    1.733125E-03    2.914500E-05
-1024   1000    9.652240E-04    2.579194E-03    3.444784E-03    3.385600E-05
-2048   1000    1.049648E-03    5.528931E-03    6.991116E-03    1.403042E-03
-4096   1000    1.774110E-03    1.143617E-02    1.418569E-02    2.414200E-03
-8192   1000    2.192489E-03    2.279767E-02    2.892733E-02    4.226552E-03
-16384  1000    2.619589E-03    4.554571E-02    5.601375E-02    8.117745E-03
-Time taken by random malloc-traversal-free benchmark with following commutative distribution: malloc 1, write 5, read 9, free 10, max malloc size 16384, length of table 1000, number of ops 1000 is 3.968187E-02.
+1  1000    2.867600E-05    8.450000E-07    1.407000E-06    1.690900E-05
+2  1000    2.900700E-05    3.020000E-06    4.845000E-06    1.677900E-05
+4  1000    2.801900E-05    6.082000E-06    1.125100E-05    1.672700E-05
+8  1000    2.845000E-05    1.308800E-05    2.378200E-05    1.678200E-05
+16 1000    4.802200E-05    2.919800E-05    4.941700E-05    2.274500E-05
+32 1000    4.469200E-05    7.263600E-05    1.076410E-04    2.315000E-05
+64 1000    9.601300E-05    1.505090E-04    2.069160E-04    2.355300E-05
+128    1000    1.647250E-04    2.961200E-04    4.069040E-04    2.409200E-05
+256    1000    3.110510E-04    5.955610E-04    8.107640E-04    2.445600E-05
+512    1000    7.043500E-04    2.086689E-02    1.613696E-03    2.687500E-05
+1024   1000    8.665660E-04    2.507518E-03    3.235066E-03    3.006300E-05
+2048   1000    2.195796E-02    5.197294E-03    2.654418E-02    9.783890E-04
+4096   1000    3.094639E-02    5.448493E-02    5.318073E-02    1.760589E-03
+8192   1000    8.852100E-03    7.927022E-02    1.085874E-01    3.304151E-03
+16384  1000    2.596275E-02    1.620433E-01    2.173120E-01    2.720816E-02
+Time taken by random malloc-traversal-free benchmark with following commutative distribution: malloc 1, write 5, read 9, free 10, max malloc size 16384, length of table 1000, number of ops 1000 is 1.464438E-01.
 [overhead] initiated
-[overhead] (CmiSyncSend) 1.810361e-07 seconds per 0 bytes
-[overhead] (CmiSyncSend) 1.792061e-07 seconds per 16 bytes
-[overhead] (CmiSyncSend) 2.133896e-07 seconds per 256 bytes
-[overhead] (CmiSyncSend) 2.802949e-07 seconds per 4096 bytes
-[overhead] (CmiSyncSend) 2.981894e-06 seconds per 65536 bytes
-[overhead] (CmiSyncSend) 1.141689e-04 seconds per 1048576 bytes
-[overhead] (CmiSyncSendAndFree) 1.198125e-07 seconds per 0 bytes
-[overhead] (CmiSyncSendAndFree) 1.179336e-07 seconds per 16 bytes
-[overhead] (CmiSyncSendAndFree) 1.268184e-07 seconds per 256 bytes
-[overhead] (CmiSyncSendAndFree) 1.140879e-07 seconds per 4096 bytes
-[overhead] (CmiSyncSendAndFree) 1.141963e-07 seconds per 65536 bytes
-[overhead] (CmiSyncSendAndFree) 1.149404e-07 seconds per 1048576 bytes
-[overhead] (CsdEnqueue) 1.221953e-07 seconds per 0 bytes
-[overhead] (CsdEnqueue) 1.207207e-07 seconds per 16 bytes
-[overhead] (CsdEnqueue) 1.245771e-07 seconds per 256 bytes
-[overhead] (CsdEnqueue) 1.291768e-07 seconds per 4096 bytes
-[overhead] (CsdEnqueue) 1.203682e-07 seconds per 65536 bytes
-[overhead] (CsdEnqueue) 1.360615e-07 seconds per 1048576 bytes
+[overhead] (CmiSyncSend) 1.634209e-07 seconds per 0 bytes
+[overhead] (CmiSyncSend) 1.604414e-07 seconds per 16 bytes
+[overhead] (CmiSyncSend) 1.938555e-07 seconds per 256 bytes
+[overhead] (CmiSyncSend) 2.647725e-07 seconds per 4096 bytes
+[overhead] (CmiSyncSend) 2.819311e-06 seconds per 65536 bytes
+[overhead] (CmiSyncSend) 4.474338e-04 seconds per 1048576 bytes
+[overhead] (CmiSyncSendAndFree) 1.059512e-07 seconds per 0 bytes
+[overhead] (CmiSyncSendAndFree) 1.039082e-07 seconds per 16 bytes
+[overhead] (CmiSyncSendAndFree) 1.039355e-07 seconds per 256 bytes
+[overhead] (CmiSyncSendAndFree) 1.038945e-07 seconds per 4096 bytes
+[overhead] (CmiSyncSendAndFree) 1.039238e-07 seconds per 65536 bytes
+[overhead] (CmiSyncSendAndFree) 1.042666e-07 seconds per 1048576 bytes
+[overhead] (CsdEnqueue) 1.150664e-07 seconds per 0 bytes
+[overhead] (CsdEnqueue) 1.116953e-07 seconds per 16 bytes
+[overhead] (CsdEnqueue) 1.112529e-07 seconds per 256 bytes
+[overhead] (CsdEnqueue) 1.116436e-07 seconds per 4096 bytes
+[overhead] (CsdEnqueue) 1.120566e-07 seconds per 65536 bytes
+[overhead] (CsdEnqueue) 1.117246e-07 seconds per 1048576 bytes
 [timer] initiated
-[timer] (CmiCpuTimer) 5.210627e-07 seconds per call
-[timer] (CmiWallTimer) 2.870076e-08 seconds per call
-[timer] (CmiCpuTimer) 2.000000e-06 seconds resolution
-[timer] (CmiWallTimer) 9.500000e-08 seconds resolution
+[timer] (CmiCpuTimer) 2.149742e-06 seconds per call
+[timer] (CmiWallTimer) 8.535737e-08 seconds per call
+[timer] (CmiCpuTimer) 1.000000e-06 seconds resolution
+[timer] (CmiWallTimer) 8.800000e-08 seconds resolution
 [proc] initiated
 proc init
 if
@@ -84,47 +84,47 @@
 else
 else
 else
-[proc] Tseq = 1.328615e-01 seconds, Tpar = 3.341508e-02 seconds
+[proc] Tseq = 4.831276e-01 seconds, Tpar = 1.598084e-01 seconds
 [proc] CmiNumPes() reported 4 processors
-[proc] But actual number of processors is 4
+[proc] But actual number of processors is 3
 [proc] FYI, appox PI (seq) = 3.141210
 [proc] FYI, appox PI (par) = 3.142746
 else if
 [smputil] initiated
-[smputil] StackVar Access Overhead: 9.634461e-10 seconds
-[smputil] ProcPrivateVar Access Overhead: 9.585953e-10 seconds
-[smputil] SharedVar Access Overhead: 9.566697e-10 seconds
-[smputil] LockUnlock Overhead: 2.644859e-09 seconds
-[smputil] Estimated CmiAlloc/CmiFree Overhead (w/o contention): 1.948589e-07 seconds
-[smputil] Estimated CmiAlloc/CmiFree Overhead (w contention): 2.279238e-07 seconds
-[smputil] Barrier Overhead: -9.050000e-11 seconds
+[smputil] StackVar Access Overhead: 4.181194e-09 seconds
+[smputil] ProcPrivateVar Access Overhead: 3.193258e-09 seconds
+[smputil] SharedVar Access Overhead: 1.502988e-09 seconds
+[smputil] LockUnlock Overhead: 2.424347e-09 seconds
+[smputil] Estimated CmiAlloc/CmiFree Overhead (w/o contention): 5.152627e-07 seconds
+[smputil] Estimated CmiAlloc/CmiFree Overhead (w contention): 6.730922e-07 seconds
+[smputil] Barrier Overhead: -4.300000e-12 seconds
 [pingpong] initiated
 [pingpong] CmiSyncSend
-[pingpong] size=16 averageTime=9.363042e-06 seconds
-[pingpong] size=16 maxTime=2.886200e-05 seconds    [1->2]
-[pingpong] size=16 minTime=2.224500e-06 seconds    [2->1]
-[pingpong] size=128    averageTime=2.238557e-06 seconds
-[pingpong] size=128    maxTime=5.193226e-06 seconds    [1->3]
-[pingpong] size=128    minTime=6.767325e-07 seconds    [2->0]
-[pingpong] size=512    averageTime=1.809700e-06 seconds
-[pingpong] size=512    maxTime=2.844870e-06 seconds    [0->3]
-[pingpong] size=512    minTime=7.852800e-07 seconds    [2->1]
-[pingpong] size=2048   averageTime=2.639384e-06 seconds
-[pingpong] size=2048   maxTime=4.035995e-06 seconds    [0->3]
-[pingpong] size=2048   minTime=1.079595e-06 seconds    [2->1]
-[pingpong] size=8192   averageTime=1.139257e-05 seconds
-[pingpong] size=8192   maxTime=3.111226e-05 seconds    [0->3]
-[pingpong] size=8192   minTime=2.422210e-06 seconds    [2->1]
-[pingpong] size=32768  averageTime=1.036710e-05 seconds
-[pingpong] size=32768  maxTime=1.876734e-05 seconds    [0->1]
-[pingpong] size=32768  minTime=6.475250e-06 seconds    [1->0]
-[pingpong] size=131072 averageTime=2.591648e-05 seconds
-[pingpong] size=131072 maxTime=3.495582e-05 seconds    [0->3]
-[pingpong] size=131072 minTime=2.039378e-05 seconds    [0->1]
-[pingpong] size=1048576    averageTime=1.768574e-04 seconds
-[pingpong] size=1048576    maxTime=2.046254e-04 seconds    [0->3]
-[pingpong] size=1048576    minTime=1.530103e-04 seconds    [1->0]
-[pingpong] size=8388608    averageTime=1.785023e-03 seconds
-[pingpong] size=8388608    maxTime=2.387379e-03 seconds    [0->3]
-[pingpong] size=8388608    minTime=1.491708e-03 seconds    [1->0]
+[pingpong] size=16 averageTime=2.598936e-02 seconds
+[pingpong] size=16 maxTime=3.093020e-02 seconds    [0->3]
+[pingpong] size=16 minTime=2.207381e-02 seconds    [3->0]
+[pingpong] size=128    averageTime=1.395744e-02 seconds
+[pingpong] size=128    maxTime=1.449006e-02 seconds    [3->2]
+[pingpong] size=128    minTime=1.370125e-02 seconds    [2->3]
+[pingpong] size=512    averageTime=1.407342e-02 seconds
+[pingpong] size=512    maxTime=1.469003e-02 seconds    [0->2]
+[pingpong] size=512    minTime=1.332164e-02 seconds    [2->3]
+[pingpong] size=2048   averageTime=1.410294e-02 seconds
+[pingpong] size=2048   maxTime=1.479005e-02 seconds    [0->3]
+[pingpong] size=2048   minTime=1.369444e-02 seconds    [1->0]
+[pingpong] size=8192   averageTime=1.414839e-02 seconds
+[pingpong] size=8192   maxTime=1.490997e-02 seconds    [1->3]
+[pingpong] size=8192   minTime=1.380386e-02 seconds    [2->0]
+[pingpong] size=32768  averageTime=1.404905e-02 seconds
+[pingpong] size=32768  maxTime=1.446495e-02 seconds    [3->0]
+[pingpong] size=32768  minTime=1.372995e-02 seconds    [1->0]
+[pingpong] size=131072 averageTime=1.454551e-02 seconds
+[pingpong] size=131072 maxTime=1.637387e-02 seconds    [2->0]
+[pingpong] size=131072 minTime=1.350686e-02 seconds    [0->2]
+[pingpong] size=1048576    averageTime=1.528689e-02 seconds
+[pingpong] size=1048576    maxTime=1.598968e-02 seconds    [0->3]
+[pingpong] size=1048576    minTime=1.390215e-02 seconds    [0->1]
+[pingpong] size=8388608    averageTime=1.694444e-02 seconds
+[pingpong] size=8388608    maxTime=2.144149e-02 seconds    [0->2]
+[pingpong] size=8388608    minTime=1.508169e-02 seconds    [1->2]
 [pingpong-cachemiss] initiated
```

**Comment by nitbhat** [comment]
Thanks for the analysis @matthiasdiener. 

I ran the current charm head with 2 nodes and didn't see any particular change in the time. It is still very slow and takes 25-30 odd mins. 

I also ran the previous version of charm (from sept 16, 2020 with Commit ID: dc40786), and it also is very slow and takes 25-30 odd mins. (in comparison to the 6s it took when running it back then). 

I also custom built UCX 1.9.0 and tested it with the current charm master and it also didn't change things. So, I'm guessing that this issue is mostly because of some change in probably the machine's network libraries (and not a change in charm). 

I'm already in conversation with the campus cluster's sysadmin about this issue. I'm curious as to what change over the last 4-5 months has caused this massive performance degradation.
